src/compiler/checker.ts

```diff
@@ -16729,6 +16743,24 @@ namespace ts {
             return true;
         }
 
+        function checkAndAggregateParameterExpressionTypes(parameter: ParameterDeclaration): Type[] {
+            const func = <FunctionLikeDeclaration>parameter.parent;
+            const usageTypes: Type[] = [];
+            forEachInvocation(<Block>func.body, invocation => {

```

Comment:
@HerringtonDarkholme I've put out some initial thoughts in https://github.com/Microsoft/TypeScript/issues/15114#issuecomment-321455766. We can exploit co and contravariance to get sensible rules here. If something assigns to a parameter, the sensible thing is for the parameter to be inferred as having a *supertype* of said assignment. Unfortunately TypeScript can't represent this.

This means we should restrict ourselves to inference in *covariant* usages of the parameter variable, i.e. wherever it is used as a source of information (assignment of the parameter to well-typed reference, invocation of the parameter with well-typed argument list, accessing properties on parameter, etc.)